src/lib/mina_graphql/mina_graphql.ml

```diff
@@ -4340,81 +4340,116 @@ module Queries = struct
+        let mina_config = Mina_lib.config mina in
+        let local_state = mina_config.consensus_local_state in
+        let constants = mina_config.precomputed_values.consensus_constants in
+        let cast_ledger :
+               Consensus.Data.Local_state.Snapshot.Ledger_snapshot.t
+            -> Ledger.Any_ledger.witness = function
+          | Genesis_epoch_ledger l ->
+              Ledger.Any_ledger.cast (module Ledger) l
+          | Ledger_db l ->
+              Ledger.Any_ledger.cast (module Ledger.Db) l
         in
-        assert (
-          Mina_base.Ledger_hash.equal
-            (Ledger.Any_ledger.M.merkle_root staking_ledger)
-            staking_epoch.ledger.hash ) ;
-        let%bind next_epoch_ledger =
-          match
-            (* We always want to return the next epoch ledger here, in case we
-               need to hard-fork from a block where it is unfinalized. The
-               safety concern doesn't apply here, because we are only using
-               this to build a snapshot, and never applying it back to the
-               running network.
-            *)
-            Mina_lib.next_epoch_ledger
-              ~unsafe_always_return_ledger_as_if_finalized:true mina
-          with
-          | None ->
-              Deferred.Result.fail "Next epoch ledger is not initialized."
-          | Some `Notfinalized ->
-              failwith "next_epoch_ledger returned a disallowed value"
-          | Some (`Finalized (Genesis_epoch_ledger l)) ->
-              return (Ledger.Any_ledger.cast (module Ledger) l)
-          | Some (`Finalized (Ledger_db l)) ->
-              return (Ledger.Any_ledger.cast (module Ledger.Db) l)
+        let local_cur_epoch_ledger =
+          Consensus.Data.Local_state.staking_epoch_ledger local_state
+          |> cast_ledger
+        in
+        let local_next_epoch_ledger =
+          Consensus.Data.Local_state.next_epoch_ledger local_state
+          |> cast_ledger
+        in
+        let best_tip_finalized =
+          Consensus.Hooks.epoch_ledgers_finalized ~constants
+            best_tip_consensus_state
+        in
+        let epoch =
+          Consensus.Proof_of_stake.Exported.Global_slot.(
+            Fn.compose epoch (of_slot_number ~constants))
+        in
+        let same_epoch =
+          Unsigned.UInt32.equal (epoch target_slot)
+          @@ epoch
+          @@ Consensus.Data.Consensus_state.curr_global_slot
+               best_tip_consensus_state
+        in
+        let%bind staking_ledger, next_epoch_ledger =
+          if same_epoch && not best_tip_finalized then
+            (* Neither target nor best tip are finalized (both in the same epoch) =>
+               `next_epoch_ledger` refers to staking ledger corresponding to staking ledger of
+               the current epoch (because root is in the previous epoch) *)
+            let%map ledger =
+              Mina_lib.get_snarked_ledger_full mina (Some target_state_hash)
```

Comment:
This is the incorrect snarked ledger. You want the first block after the epoch transition.